csrc/index_compute.cpp

```diff
@@ -3291,81 +3291,594 @@ Val* Index::eye(
   return GpuLower::current()->commonScalarMap().hoistScalar(result, loops);
 }
 
-Val* Index::cpAsyncBulkIndex(
-    TensorView* gmem_tv,
-    TensorView* smem_tv,
-    TensorView* consumer,
-    Val* mbarrier,
-    const std::vector<kir::ForLoop*>& loops) {
-  using namespace tma;
-
-  bool is_load = (gmem_tv != consumer);
-
-  NVF_ERROR(
-      gmem_tv->getMemoryType() == MemoryType::Global,
-      "cpAsyncBulkIndex is only for global memory tensors");
-
-  int64_t dim = (int64_t)gmem_tv->getMaybeAllocationDomain().size();
-  NVF_ERROR(dim > 0);
-  int64_t itemsize = dataTypeSize(gmem_tv->dtype());
+// Note [Schedule and indexing of TMA]
+//
+// TMA is a hardware feature that allows the GPU to load a tile from a tensor of
+// up to 5D. Note that the dimensionality of the tensor in the eyes of TMA is
+// not necessarily the same as the dimensionality of the tensor in the eyes of
+// the user. For example, it is possible that there is a 1D tensor of shape
+// 1000000 and the user wants to view it as (1000, 1000) and load a tile of (8,
+// 8) from it. Another example is, the user has a tensor of shape (1000, 1000),
+// and as long as the tensor is contiguous, the user can view it as a flattened
+// shape 1000000 tensor, and load a tile of 8 from it. The dimensionality of the
+// tensor in the eyes of TMA is determined by the schedule of the tensor.
+// However, the dimensionality of the tensor in the eyes of TMA can not be
+// smaller than the physical dimensionality of the tensor. For example, if the
+// user has a tensor of shape (128, 128, 128, 128, 128, 128), and the tensor is
+// fully discontiguous, then it is not possible to view it as a <= 5D tensor,
+// thus TMA can not be used to load/store a tile from it.
+//
+// The gmem tensor of TMA store, or the gmem tensor of TMA load when replayed as
+// consumer, must be scheduled like this:
+//
+//                         ---------------------
+//                         | Allocation Domain |
+//                         ---------------------
+//                           | | | | | | | | |
+//                         IterDomain Expressions
+//                             | | | | | | |
+//                         ---------------------
+//                         |  I1           I2  |
```

Comment:
Should we define this part of the domains, i.e., `{I1, I2}`, as the TMA domain? Then, we could define the box domains as the inner split output domains of the TMA domain.

IIUC, the global domains defined below is the same as the TMA domain, right?

Reply:
> IIUC, the global domains defined below is the same as the TMA domain, right?

Yes

> Should we define this part of the domains, i.e., `{I1, I2}`, as the TMA domain? Then, we could define the box domains as the inner split output domains of the TMA domain.

I do think we should give `{I1, I2}` a name, and TMA domain do sound like a good name. But conceptually, is it possible to **define** the box domain as a split of TMA domain? I think we can only **understand** the box domain as a split of TMA domain, but not **define** it this way. We need to define "global ID" first, because otherwise, how would you describe what IDs are included in TMA domain?

Reply:
I just meant to say we name the split domain of a TMA domain as the box domain. Yes, we would first need to identify what domains should be included in the TMA domain.

Reply:
updated in https://github.com/NVIDIA/Fuser/pull/1962/commits/62dd8b9fb72872b21336b4db128892c7fb6dae1e